### ISSUE TITLE
WASM: Ungraceful disconnections don't despawn entities or send the `DisconnectEvent`

### ISSUE THREAD

**Issue by simbleau** [cBournhonesque/lightyear#488]
If you adjust the simple_box example to use the `handle_disconnections` system, e.g.

```rust
...
    app.add_systems(Update, handle_disconnections);
...

pub(crate) fn handle_disconnections(
    mut disconnections: EventReader<DisconnectEvent>,
) {
    for disconnection in disconnections.read() {
        error!("Client {:?} disconnected", disconnection.client_id);
    }
}
```

Then close the tab, browser, etc. The server will not receive the event, it is never sent.

**Comment by simbleau** [comment]
I'm at a wall because my clients entities are not being removed when they disconnect (which is advertised as the default) but I can't seem to use the `DisconnectEvent` to do it myself either.

**Comment by simbleau** [comment]
edit: I narrowed it down a bit. The event only fails to be sent when the user is closing the tab or having an otherwise ungraceful disconnection.

**Comment by cBournhonesque** [comment]
> I'm at a wall because my clients entities are not being removed when they disconnect (which is advertised as the default) but I can't seem to use the DisconnectEvent to do it myself either.

The entities are only despawned automatically if you added a `ControlledBy` entity on the server, I assume you did that?

Let me check if indeed DisconnectEvent is not emitted from io-based timeouts

**Comment by cBournhonesque** [comment]
I'm not able to reproduce this; if I close a tab the entities still get despawned on the server after 3 seconds

**Comment by simbleau** [comment]
Haven't been at the PC, but yes. Every entity that isn't being despawned has a "ControlledBy".

I copied the Replicate settings from the simple_box.

**Comment by simbleau** [comment]
Here's been my debugging, roughly.

On the left we have the server inspector, showing the world at runtime.

When a client connects, I see the `ControlledBy` component attached to the player.

<img width="1553" alt="image" src="https://github.com/cBournhonesque/lightyear/assets/48108917/13af1784-8fd1-47cc-adfc-efbeb3159268">

When they disconnect, here's what the server logs look like, with debug on:

```
2024-06-26T15:07:11.341323Z DEBUG lightyear::shared::replication::send: Sending replication messages messages=[(ChannelKind(TypeId { t: (7545640365327152828, 651029402225809427) }), ReplicationGroupId(4294967297), Updates(EntityUpdatesMessage { last_action_tick: Some(Tick(499)), updates: [(1v1, [[4, 0, 190, 86, 18, 196, 83, 116, 144, 67, 219, 15, 201, 191, 0]])] }), 1656.0)]
2024-06-26T15:07:11.342768Z  INFO lightyear::transport::webtransport::server: Connection with 127.0.0.1:63527 closed. Reason: TimedOut
2024-06-26T15:07:11.342789Z DEBUG lightyear::transport::webtransport::server: Dropping tasks
2024-06-26T15:07:11.342821Z DEBUG Driver{quic_id=5729550864}: wtransport::driver::worker: Ended with error: NotConnected
2024-06-26T15:07:11.355274Z ERROR lightyear::server::networking: Disconnect client 127.0.0.1:63527 because io task failed
2024-06-26T15:07:11.439333Z DEBUG lightyear::server::input: input buffer for client input_buffer=InputBuffer { buffer: [], start_tick: Some(Tick(2160)) } tick=Tick(2160) client_id=Netcode(13174697591107995770)
2024-06-26T15:07:11.439376Z DEBUG lightyear::server::input: Missed client input! client_id=Netcode(13174697591107995770) tick=Tick(2160) fallback_input=Some(Standing { at: Vec2(-585.35535, 288.90878), facing: Direction(-1.5707964) })
2024-06-26T15:07:11.440975Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=0 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.441076Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=1 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.441142Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=2 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.441204Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=4 change_tick=Tick { tick: 4296188 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.441561Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=3 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.441752Z DEBUG lightyear::shared::replication::send: Sending replication messages messages=[(ChannelKind(TypeId { t: (7545640365327152828, 651029402225809427) }), ReplicationGroupId(4294967297), Updates(EntityUpdatesMessage { last_action_tick: Some(Tick(499)), updates: [(1v1, [[4, 0, 190, 86, 18, 196, 83, 116, 144, 67, 219, 15, 201, 191, 0]])] }), 1657.0)]
2024-06-26T15:07:11.538982Z DEBUG lightyear::server::input: input buffer for client input_buffer=InputBuffer { buffer: [], start_tick: Some(Tick(2161)) } tick=Tick(2161) client_id=Netcode(13174697591107995770)
2024-06-26T15:07:11.539022Z DEBUG lightyear::server::input: Missed client input! client_id=Netcode(13174697591107995770) tick=Tick(2161) fallback_input=Some(Standing { at: Vec2(-585.35535, 288.90878), facing: Direction(-1.5707964) })
2024-06-26T15:07:11.540505Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=0 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.540598Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=1 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.540674Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=2 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.540785Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=4 change_tick=Tick { tick: 4297577 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.541084Z DEBUG lightyear::server::connection: prepare entity update changed check (we want the component-change-tick to be higher than collect-changes-since-this-tick) kind=3 change_tick=Tick { tick: 740124 } collect_changes_since_this_tick=Some(Tick { tick: 4153151 })
2024-06-26T15:07:11.541247Z DEBUG lightyear::shared::replication::send: Sending replication messages messages=[(ChannelKind(TypeId { t: (7545640365327152828, 651029402225809427) }), ReplicationGroupId(4294967297), Updates(EntityUpdatesMessage { last_action_tick: Some(Tick(499)), updates: [(1v1, [[4, 0, 190, 86, 18, 196, 83, 116, 144, 67, 219, 15, 201, 191, 0]])] }), 1658.0)]
2024-06-26T15:07:11.639218Z DEBUG lightyear::server::input: input buffer for client input_buffer=InputBuffer { buffer: [], start_tick: Some(Tick(2162)) } tick=Tick(2162) client_id=Netcode(13174697591107995770)
2024-06-26T15:07:11.639257Z DEBUG lightyear::server::input: Missed client input! client_id=Netcode(13174697591107995770) tick=Tick(2162) fallback_input=Some(Standing { at: Vec2(-585.35535, 288.90878), facing: Direction(-1.5707964) })
```

5 minutes later, it's still trying to send replication messages and thinks the client is connected, but I clearly closed the tab and ended the connection.

<img width="1049" alt="image" src="https://github.com/cBournhonesque/lightyear/assets/48108917/77a332fb-00f0-4744-bb11-e0beb691f51c">

**Comment by cBournhonesque** [comment]
Hm are you using the released version 0.15.1?
I think this is fixed in main, which is why I cannot reproduce it

**Comment by simbleau** [comment]
Yes, I'm using 0.15.1

**Comment by simbleau** [comment]
When is the next release planned?

**Comment by cBournhonesque** [comment]
As soon as bevy 0.14.0 launches, which should be this weekend. The release will be by the end of the following week (since I have to wait for leafwing-input-manager, xpbd, etc. to also upgrade)

**Comment by simbleau** [comment]
Ah, ok. Let me see if I can reproduce this on main then...

**Comment by simbleau** [comment]
Update: Works on `main`, so I suppose I can close this issue now.

**Comment by simbleau** [comment]
Edit: on main, after I disconnect, it still attempts to process inputs, I thought this was fixed?

<img width="1613" alt="image" src="https://github.com/cBournhonesque/lightyear/assets/48108917/466c24ed-cdb5-478f-97d2-3286f31efdb7">

**Comment by Nul-led** [comment]
did u cargo update to latest commit?

**Comment by simbleau** [comment]
Yep

**Comment by simbleau** [comment]
I am using the latest commit currently, 

```rust
lightyear = { git = "https://github.com/cBournhonesque/lightyear", rev = "10d7a9f", features = [
  "webtransport",
] }
```

**Comment by cBournhonesque** [comment]
It was fixed on the branch `cb/0.16` (which works with bevy 0.14) but not on main. 
Sorry it's a bit confusing, but I've started pushing most fixes to `cb/0.16` because there's a number of changes there that make it easier.

**Comment by cBournhonesque** [comment]
But great to know that the ungraceful disconnections work for you on `main` !
